src/rdrand.rs

```diff
@@ -9,33 +9,57 @@
 //! Implementation for SGX using RDRAND instruction
 use crate::Error;
 use core::mem;
-use core::arch::x86_64::_rdrand64_step;
+use core::arch::x86_64::{__cpuid, _rdrand64_step};
 use core::num::NonZeroU32;
-
-#[cfg(not(target_feature = "rdrand"))]
-compile_error!("enable rdrand target feature!");
+use lazy_static::lazy_static;
 
 // Recommendation from "Intel® Digital Random Number Generator (DRNG) Software
 // Implementation Guide" - Section 5.2.1 and "Intel® 64 and IA-32 Architectures
 // Software Developer’s Manual" - Volume 1 - Section 7.3.17.1.
 const RETRY_LIMIT: usize = 10;
 const WORD_SIZE: usize = mem::size_of::<u64>();
 
-fn rdrand() -> Result<[u8; WORD_SIZE], Error> {
+#[target_feature(enable = "rdrand")]
+unsafe fn rdrand() -> Result<[u8; WORD_SIZE], Error> {
     for _ in 0..RETRY_LIMIT {
-        unsafe {
-            // SAFETY: we've checked RDRAND support, and u64 can have any value.
-            let mut el = mem::uninitialized();
-            if _rdrand64_step(&mut el) == 1 {
-                return Ok(el.to_ne_bytes());
-            }
-        };
+        let mut el = mem::uninitialized();
+        if _rdrand64_step(&mut el) == 1 {
+            return Ok(el.to_ne_bytes());
+        }
     }
     error!("RDRAND failed, CPU issue likely");
     Err(Error::UNKNOWN)
 }
 
+// TODO use is_x86_feature_detected!("rdrand") when that works in core. See:
+//   https://github.com/rust-lang-nursery/stdsimd/issues/464
+fn is_rdrand_supported() -> bool {
+    if cfg!(target_feature = "rdrand") {
+        true
+    } else if cfg!(target_env = "sgx") {
+        false // No CPUID in SGX enclaves
```

Comment:
Shouldn't we issue a compilation error here?

Reply:
I'm just copying what's in `std_detect`, and I don't think we should deviate from their implementation.

Reply:
Also, if we were going to deviate from `std_detect`, it would probably be to just always assume RDRAND is present on SGX, see https://github.com/nagisa/rust_rdrand/issues/9. But I don't think that is a good idea either.

Reply:
I mean why not just use:
```rust
#[cfg(not(and(target_feature = "rdrand", target_env = "sgx")))]
compile_error!("..");
```
and drop the SGX branch. If somehow code which targets SGX gets compile without `rdrand` feature (considering that it's enabled by default such event should be quite unlikely), developer will get a compilation error instead of always failing application.

Reply:
Added a check for SGX && !RDRAND

Reply:
Ups, you should use `all` instead of `and`.